R/utilities.R

```diff
@@ -2646,3 +2646,267 @@ standardize_chr_prefix = function(incoming_vector,
+  if (projection == "grch37") {
+    chr_coordinates = chromosome_arms_grch37
+  } else if (projection == "hg38") {
+    chr_coordinates = chromosome_arms_hg38
+  } else {
+    stop(
+      "You specified projection that is currently not supported. Please provide seg files in either hg38 or grch37."
+    )
+  }
+
+  # exclude sex chromosomes
+  if (exclude_sex) {
+    chr_coordinates = chr_coordinates %>%
+      dplyr::filter(!grepl("X|Y", chromosome))
+  }
+
+  # does the user's seg file contain centromeres?
+  if (exclude_centromeres) {
+    chr_coordinates = chr_coordinates %>%
+      group_by(chromosome) %>%
+      mutate(start = min(start),
+             end = max(end)) %>%
+      ungroup
+  }
+
+  # total size of genome in this projection
+  genome_size = chr_coordinates %>%
+    mutate(size = end - start) %>%
+    summarise(genome_size = sum(size)) %>%
+    pull(genome_size)
+
+  # prepare for the overlaps
+  chr_coordinates = as.data.table(chr_coordinates)  %>%
+    rename("arm_start" = "start",
+           "arm_end" = "end")
+  setkey(chr_coordinates, chromosome, arm_start, arm_end)
+
+  # work out the seg file
+  if (!missing(seg_path)) {
+    message(paste0("Reading thhe seg file from ", seg_path))
+    this_seg = read_tsv(seg_path)
+  }
+
+  # preserve the sample ids to account later for those with 0 PGA
+  sample_set = this_seg %>% distinct(sample)
+
+  this_seg = this_seg %>%
+    dplyr::filter(abs(log.ratio) >= cutoff) %>%
+    dplyr::relocate(sample, .after = last_col())
+
+  # ensure consistent chromosome prefixing
+  if (projection == "grch37") {
+    this_seg$chrom = gsub("chr", "", this_seg$chrom)
+  } else {
+    this_seg$chrom = gsub("chr", "", this_seg$chrom) # if there is a mish-mash of prefixes, strip them all
+    this_seg$chrom = paste0("chr", this_seg$chrom)
+  }
+
+  # exclude sex chromosomes
+  if (exclude_sex) {
+    this_seg = this_seg %>%
+      dplyr::filter(!grepl("X|Y", chrom))
+  }
+
+  # prepare for the overlaps
+  this_seg = as.data.table(this_seg)
+  setkey(this_seg, chrom, start, end)
+
+  # what are the segments that overlap good regions in chromosome coordinates?
+  this_seg = foverlaps(
+    this_seg,
+    chr_coordinates,
+    by.x = c("chrom", "start", "end"),
+    by.y = c('chromosome', 'arm_start', 'arm_end'),
+    nomatch = 0L
+  ) %>%
+    as.data.frame %>%
+    arrange(sample, chrom, start)
+
+  # calculate total length of CNV
+  affected_regions = this_seg %>%
+    dplyr::mutate(size = end - start) %>%
+    group_by(sample) %>%
+    summarise(total = sum(size))
+
+  affected_regions$PGA = affected_regions$total / genome_size
+
+  # now add any samples that can have 0 PGA
+  affected_regions = base::merge(sample_set,
+                                 affected_regions,
+                                 all.x = TRUE) %>%
+    replace_na(list(PGA = 0))
+
+  affected_regions = affected_regions %>%
+    select(-total) %>%
+    `names<-`(c("sample_id", "PGA")) %>%
```

Comment:
How does this work?

Reply:
I specify the function within the tickmarks and then in parentheses what should go there. It can also recognize other functions like `rownames` etc - a good approach to renaming columns/rows in pipe. So here I just specified the column names